stdlib/public/runtime/Casting.cpp

```diff
@@ -2876,8 +2876,50 @@ id _bridgeAnythingNonVerbatimToObjectiveC(OpaqueValue *src,
 extern "C" const _ObjectiveCBridgeableWitnessTable BRIDGING_CONFORMANCE_SYM;
 #endif
 
+struct ObjCBridgeConformanceMemo {
+  // TODO: If asserts are not being compiled, don't allocate this extra var
+  const Metadata *destType;
+  const _ObjectiveCBridgeableWitnessTable *destBridgeWitness;
+  swift_once_t fetchWitnessOnce;
+
+  const _ObjectiveCBridgeableWitnessTable *
+  findBridgeWitness(const Metadata *T) {
+    struct SetupData {
+      const Metadata *destType;
+      struct ObjCBridgeConformanceMemo *memo;
+    } setupData { T, this };
+
+    swift_once(&fetchWitnessOnce,
+               [](void *data) {
+                 struct SetupData *setupData = (struct SetupData *)data;
+                 struct ObjCBridgeConformanceMemo *memo = setupData->memo;
+                 // Check that this always gets called with the same destType.
+                 assert((memo->destType == nullptr) || (memo->destType == setupData->destType));
+                 memo->destType = setupData->destType;
+                 auto w = swift_conformsToProtocol(
+                   memo->destType, &PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable));
+                 memo->destBridgeWitness =
+                   reinterpret_cast<const _ObjectiveCBridgeableWitnessTable *>(w);
+               }, (void *)&setupData);
+    return destBridgeWitness;
+  }
+};
+
+/// Nominal type descriptor for Swift.String.
+extern "C" const StructDescriptor NOMINAL_TYPE_DESCR_SYM(SS);
+
 static const _ObjectiveCBridgeableWitnessTable *
 findBridgeWitness(const Metadata *T) {
+  // Special case: Memoize the bridge witness for Swift.String.
+  // Swift.String is the most heavily used bridge because of the prevalence of
+  // string-keyed dictionaries in Obj-C.  It's worth burning a few words of static
+  // storage to avoid repeatedly looking up this conformance.
+  if (T->getKind() == MetadataKind::Struct
+      && (cast<StructMetadata>(T)->Description == &NOMINAL_TYPE_DESCR_SYM(SS))) {
+    static ObjCBridgeConformanceMemo memo;
```

Comment:
This is c++. Can't you put the computation of the bridging witness into the initializer. c++11 says that should be thread safe, no?

Reply:
Specifically: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

Reply:
Do we have any guarantees that the static initializer won't run until after dynamic libraries get loaded?  (In particular, the Foundation overlay.)

Reply:
Thanks for the pointer!  I've reworked the code along these lines.  Let's see what the benchmark says about it.